File: lesson_13/maps_java/maps_app/src/main/java/com/codedifferently/lesson13/Lesson13.java
package com.codedifferently.lesson13;

import java.util.HashMap;
import java.util.Map;

public class Lesson13 {

  /**
   * Provide the solution to LeetCode 3146 here:
   * https://leetcode.com/problems/permutation-difference-between-two-strings
   */
  public int findPermutationDifference(String s, String t) {
    Map<Character, Integer> sIndexMap = new HashMap<>();
    Map<Character, Integer> tIndexMap = new HashMap<>();

    // Store the index of each character in string s
    for (int i = 0; i < s.length(); i++) {
      sIndexMap.put(s.charAt(i), i);
    }

    // Store the index of each character in string t
    for (int i = 0; i < t.length(); i++) {
      tIndexMap.put(t.charAt(i), i);
    }

    int difference = 0;

    // Calculate the sum of absolute differences between indices
    for (char c : sIndexMap.keySet()) {
      int sIndex = sIndexMap.get(c);
      int tIndex = tIndexMap.get(c);
      difference += Math.abs(sIndex - tIndex);
    }

    return difference;
  }
}

File: lesson_13/maps_ts/src/lesson13.ts
/**
 * Provide the solution to LeetCode 3146 here:
 * https://leetcode.com/problems/permutation-difference-between-two-strings
 */
export function findPermutationDifference(s: string, t: string): number {
  const sIndexMap: Record<string, number> = {};
  const tIndexMap: Record<string, number> = {};

  for (let i = 0; i < s.length; i++) {
    sIndexMap[s[i]] = i;
  }

  for (let i = 0; i < t.length; i++) {
    tIndexMap[t[i]] = i;
  }

  // Sum of absolute differences between indices
  let total = 0;
  for (const char in sIndexMap) {
    const sIndex = sIndexMap[char];
    const tIndex = tIndexMap[char];
    const diff = sIndex - tIndex;
    total += diff >= 0 ? diff : -diff; 
  }

  return total;
}
